File: day9/valid_palindrome.cpp
#include<string>

using namespace std;

class Solution {
public:
    bool isPalindrome(string s) {
        int i = 0;
        int j = s.length()-1;

        while( j > i ){
            while ( !isalnum(s[i]) && j > i ){
                ++i;
            }
            while ( !isalnum(s[j]) && j > i ){
                --j;
            }

            if( tolower(s[i]) != tolower(s[j]) ){
                return false;
            }

            ++i;
            --j;
        }

        return true;
    }
};

File: solutions/day_009/valid_palindrome.java
import java.util.regex.Matcher;
import java.util.regex.Pattern;

class Solution {
  public boolean isPalindrome(String s) {
    // 1. Remove all non-alphanumeric characters
    Pattern pattern = Pattern.compile("[^a-z0-9]");
    // 2. Convert to lower case
    Matcher matcher = pattern.matcher(s.toLowerCase());
    // 3. Remove all non-alphanumeric characters
    String processedString = matcher.replaceAll("");
    // 4. Check if the string is a palindrome
    return processedString.equals(new StringBuilder(processedString).reverse().toString());
  }
}